eng/common/scripts/get-codeowners.ps1

```diff
@@ -1,49 +1,94 @@
-  exit 1
-}
+$ToolCommandName = "retrieve-codeowners"
 
-$codeOwnersContent = Get-Content $codeOwnersLocation
+function Get-CodeOwnersTool()
+{
+  # Check if the retrieve-codeowners tool exsits or not.
+  if (Get-Command "$ToolPath/$ToolCommandName" -errorAction SilentlyContinue) {
+    return "$ToolPath/$ToolCommandName"
+  }
+  if (!(Test-Path $ToolPath)) {
+    New-Item -ItemType Directory -Path $ToolPath | Out-Null
+  }
+  Write-Warning "Installing the ToolCommandName tool under $ToolPath... "
+  dotnet tool install --tool-path $ToolPath --add-source $DevOpsFeed --version $ToolVersion "Azure.Sdk.Tools.RetrieveCodeOwners" | Out-Null
 
-foreach ($contentLine in $codeOwnersContent) {
-  if (-not $contentLine.StartsWith("#") -and $contentLine){
-    $splitLine = $contentLine -split "\s+"
-    
-    # CODEOWNERS file can also have labels present after the owner aliases
-    # gh aliases start with @ in codeowners. don't pass on to API calls
-    $ownedFolders[$splitLine[0].ToLower().Trim("/")] = ($splitLine[1..$($splitLine.Length)] `
-      | ? { $_.StartsWith("@") } `
-      | % { return $_.substring(1) }) -join ","
+  $command = Join-Path $ToolPath $ToolCommandName
+  # Test to see if the tool properly installed.
+  if (!(Get-Command $command -errorAction SilentlyContinue)) {
+    Write-Error "The retrieve-codeowners tool is not properly installed. Please check your tool path. $ToolPath"
+    return 
   }
+  return $command
 }
 
-$results = $ownedFolders[$target]
-
-if ($results) {
-  Write-Host "Found a folder $results to match $target"
+function Get-CodeOwners ([string] $command)
+{
+  if (!$command) {
+    return @()
+  }
+  # Params $RootDirectory is already in use in cpp release pipeline. 
+  # Will use $CodeOwnerFileLocation and deprecate $RootDirectory once it is ready to retire $RootDirectory.
+  if ($RootDirectory -and !(Test-Path $CodeOwnerFileLocation)) {
+    $CodeOwnerFileLocation = Join-Path $RootDirectory ".github/CODEOWNERS"
+  }
+  
+  $codeOwnersString = & $command --target-directory $targetDirectory --code-owner-file-path $CodeOwnerFileLocation
+  # Failed at the command of fetching code owners.
+  if ($LASTEXITCODE -ne 0) {
+    return @()
+  }
+  
+  $codeOwnersJson = $codeOwnersString | ConvertFrom-Json
+  if (!$codeOwnersJson) {
+    Write-Host "No code owners returned from the path: $targetDirectory"
+    return @()
+  }
   
   if ($VsoVariable) {
-    $alreadyPresent = [System.Environment]::GetEnvironmentVariable($VsoVariable)
+    $codeOwners = $codeOwnersJson.Owners -join ","
+    Write-Host "##vso[task.setvariable variable=$VsoVariable;]$codeOwners"
+  }
+
+  return $codeOwnersJson.Owners
+}
+
 
-    if ($alreadyPresent) { 
-      $results += ",$alreadyPresent"
+function TestGetCodeOwner([string] $command) {
+  if (!$command) {
+    exit 1
+  }
+  $actualReturn = Get-CodeOwners -command $command 
+  $expectReturn = @("person1", "person2")
```

Comment:
Lets test with different target paths to ensure the fallback logic is working as expected.